torch/autograd/gradcheck.py

```diff
@@ -17,12 +17,12 @@ def zero_gradients(x):
 
 
 def make_jacobian(input, num_out):
-    if isinstance(input, torch.Tensor):
+    if isinstance(input, torch.Tensor) or hasattr(input, "__torch_function__"):
         if not input.is_floating_point() and not input.is_complex():
             return None
         if not input.requires_grad:
             return None
-        return torch.zeros(input.nelement(), num_out, dtype=input.dtype)
+        return torch.stack([torch.zeros_like(torch.reshape(input, (input.nelement(),)), dtype=input.dtype)] * num_out)
```

Comment:
What is the perf implication of this? 
This function is quite heavily used in the autograd tests and the matrix it generates here can quite big.

Reply:
Here's the slowdown for 1000x1000 outputs:

```
Time for `master`: 0.009022712707519531 s
Time for PR: 0.026513099670410156 s
Slowdown factor: 2.9384843039847794
```

<details><summary>Script used to generate</summary>

```python
import torch
import time

def zeros_1(inp, size2):
    return torch.zeros(inp.numel(), size2, dtype=inp.dtype)

def zeros_2(inp, size2):
    return torch.stack([torch.zeros_like(torch.reshape(inp, (inp.nelement(),)), dtype=inp.dtype)] * size2)

if __name__ == "__main__":
    start1 = time.time()
    inp = torch.zeros(1000, dtype=torch.float64)
    for _ in range(10):
        zeros_1(inp, 1000)
    time1 = time.time() - start1
    print(f"Time for `master`: {time1} s")

    start2 = time.time()
    inp = torch.zeros(1000, dtype=torch.float64)
    for _ in range(10):
        zeros_2(inp, 1000)
    time2 = time.time() - start2
    print(f"Time for PR: {time2} s")

    print(f"Slowdown factor: {time2 / time1}")
```
</details>

Reply:
That is not great... I am not sure what are your constraints here. But why did you need to do this change? The `zeros()` functions is not supported?

Reply:
This ties into the general issue of how to dispatch functions that don't receive a `Tensor`-like as an argument and therefore have nothing to dispatch on. There are a couple of different approaches to fix this in NumPy. See, for example, [NEP 37](https://numpy.org/neps/nep-0037-array-module.html), [NEP 35](https://numpy.org/neps/nep-0035-array-creation-dispatch-with-array-function.html).

Reply:
Ho I know the high level idea. I was asking which functions can you use/not use here?
So you can only use functions that take another Tensor as input? (so that the special dispatch can work)

Reply:
That's correct.

Reply:
I updated my comment above with another proposal, would that work for your use case?

Reply:
Can't you just use `new_zeros` here?